text/0027-foreman-plugin-forge.md

```diff
@@ -0,0 +1,141 @@
+|                                                |                            |
+|  [README.md] This section automatically reads  | [yum install plugin]       |
+|  contents of README.md and displays the docs.  |                            |
+|  README.md is the place where plugin authors   | [apt-get install plugin]   |
+|  are expected to write installation steps and  |                            |
+|  other relevant documentation.                 | [github.com/user/plugin]   |
+|                                                |                            |
++------------------------------------------------+----------------------------+
+```
+
+## Data sources
+
+Foreman plugins are implemented as Ruby Gems. Therefore, a lot of metadata is
+already available in the form of Gemspec attributes.
+
+* foreman_plugin_name
+* Plugin summary
+* Author
+
+Additionally, more metadata is available from RubyGems API
+(example:
+[foreman_remote_execution](https://rubygems.org/api/v1/gems/foreman_remote_execution.json))
+
+* Downloads
+* Downloads for current version
+* Github URL
+* Documentation URL
+
+## Web application
+
+Depending on how rich we want the Forge to be, one of the following approaches
+may be taken.
+
+### Jekyll based static Forge
+
+We're already using Jekyll for building a static website with much success. A
+Forge based on Jekyll would have:
+
+* A master plugin list in JSON/YAML format. Plugins can be added/remove from the
+  catalog by sending a PR against this file.
+
+```
+foreman_my_plugin:
+  version: 1.2
+  rubygem: <if it doesn't follow gem naming convention>
+  gitrepo: https://github.com/myusername/foreman_my_plugin
+  rpm: <if it doesn't follow rpm package naming convention>
+  deb: <if it doesn't follow deb package naming convention>
+```
+
+* An external Jekyll script would read this JSON/YAML catalog and start
+  gathering data points from RubyGems.org API. 
+  
+* Fill up templates with these data points and generate a unique static page for
+  Foreman plugin.
+
+### Rails based live Forge
+
+A Rails app would essentially do the same as above, but also allow users to
+create plugins manually. This way we can also extend it to have plugin ratings.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The biggest drawback is - are we over-engineering it? Foreman plugins are
+already distributed satisfactorily via yum/apt repos that we provide.
+
+# Alternatives
+[alternatives]: #alternatives
+
+A far simpler way to implement a Forge would be by extending Jekyll based
+theforeman.org website itself and add plugins like thus:
+https://theforeman.org/forge/foreman_my_plugin and a forge index page. Additions
+and updates to plugins can be made via PR
+to [theforeman.org](https://github.com/theforeman/theforeman.org) project.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Do we want to address plugin discoverability, or distribution too?
```

Comment:
@ohadlevy Yes, I was wondering if there can be a possibility of a single way of installing plugins. This is not a problem at the moment, but I added it for completeness.